core/include/cubos/core/gl/render_device.hpp

```diff
@@ -1454,6 +1479,94 @@ namespace cubos::core::gl
         protected:
             ShaderBindingPoint() = default;
         };
+
+        /// @brief Abstract timer.
+        class CUBOS_CORE_API Timer
+        {
+        public:
+            virtual ~Timer() = default;
+
+            /// @brief Starts a region to be timed.
+            virtual void begin() = 0;
+
+            /// @brief Ends a region.
+            virtual void end() = 0;
+
+            /// @brief Returns whether the result is available.
+            /// @return True if the result is available, false otherwise.
+            virtual bool done() = 0;
+
+            /// @brief Gets the result, in nanoseconds.
+            ///
+            /// Blocks until a result is available.
+            ///
+            /// @return Time spent in region, in nanoseconds.
+            virtual int result() = 0;
+
+        protected:
+            Timer() = default;
+        };
+
+        /// @brief Pipelined timer.
+        class CUBOS_CORE_API PipelinedTimer
+        {
+        public:
+            struct Result
+            {
+                int result;          ///< Time spent in region in the latest frame that finished rendering.
+                size_t framesBehind; ///< How many frames behind the CPU the latest rendered frame lags.
+            };
+
+            PipelinedTimer(RenderDevice* rd)
+                : mRd(rd)
+            {
+            }
+
+            ~PipelinedTimer() = default;
+
+            /// @brief Starts a region to be timed.
+            void begin()
+            {
+                if (mNeedsEnd) // make sure it doesn't begin again before ending
+                {
+                    return;
+                }
+                gl::Timer timer = mRd->createTimer();
+                timer->begin();
+                mTimers.push(timer);
+                mNeedsEnd = true;
+            }
+
+            /// @brief Ends a region.
+            /// @return Time spent in region in the latest frame that finished
+            /// rendering, in nanoseconds (-1 if no frame has finished yet),
+            /// and how many frames behind the CPU this frame lags.
+            Result end()
+            {
+                if (!mNeedsEnd) // make sure it doesn't end before beginning
+                {
+                    return {-1, static_cast<size_t>(-1)};
+                }
+
+                mTimers.back()->end();
+
+                while (!mTimers.empty() && mTimers.front()->done())
+                {
+                    mResult = mTimers.front()->result();
+                    mTimers.pop();
+                }
+
+                mNeedsEnd = false;
+
+                return {mResult, mTimers.size()};
```

Comment:
Why is `mResult` a member variable? Isn't it only used in this function?

Reply:
The idea was that it could persist after the function call, so that a valid value would still be returned in a future call if no timer had finished.

Reply:
Ah, makes sense. LGTM!